spark/v3.2/spark/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryPartitionPathsIndex.scala

```diff
@@ -0,0 +1,122 @@
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.types.StructType
+import scala.collection.mutable
+
+
+/**
+ * Implement our own in-memory index which will only list directories to avoid unnecessary file
+ * listings. Should ONLY be used to get partition directory paths. Uses table's schema to only
+ * visit partition dirs using number of partition columns depth recursively. Does NOT return files
+ * within leaf dir.
+ */
+class InMemoryPartitionPathsIndex(
+                         sparkSession: SparkSession,
+                         rootPath :Path,
+                         parameters: Map[String, String],
+                         userSpecifiedSchema: StructType,
+                         fileStatusCache: FileStatusCache = NoopCache,
+                         override val metadataOpsTimeNs: Option[Long] = None)
+  extends PartitioningAwareFileIndex(
+    sparkSession, parameters, Some(userSpecifiedSchema), fileStatusCache) {
+
+  override val rootPaths = Seq(rootPath)
+
+  @volatile private var cachedLeafFiles: mutable.LinkedHashMap[Path, FileStatus] = _
+  @volatile private var cachedLeafDirToChildrenFiles: Map[Path, Array[FileStatus]] = _
+  @volatile private var cachedPartitionSpec: PartitionSpec = _
+
+  refresh0()
+
+  override def partitionSpec(): PartitionSpec = {
+    if (cachedPartitionSpec == null) {
+        cachedPartitionSpec = inferPartitioning()
+    }
+    logTrace(s"Partition spec: $cachedPartitionSpec")
+    cachedPartitionSpec
+  }
+
+  override def allFiles(): Seq[FileStatus] = cachedLeafFiles.values.toSeq
+
+  override protected def leafFiles: mutable.LinkedHashMap[Path, FileStatus] = cachedLeafFiles
+
+  override protected def leafDirToChildrenFiles: Map[Path, Array[FileStatus]] = cachedLeafDirToChildrenFiles
+
+  override def refresh(): Unit = {
+    fileStatusCache.invalidateAll()
+    refresh0()
+  }
+
+  private def refresh0(): Unit = {
+    val files = listLeafDirs(rootPath, 0)
+    cachedLeafFiles =
+      new mutable.LinkedHashMap[Path, FileStatus]() ++= files.map(f => f.getPath -> f)
+    cachedLeafDirToChildrenFiles = files.map( f => (f.getPath, Array(createEmptyChildDataFileStatus(f))))
+      .toMap
+    cachedPartitionSpec = null
+  }
+
+  override def equals(other: Any): Boolean = other match {
+    case hdfs: InMemoryFileIndex => rootPaths.toSet == hdfs.rootPaths.toSet
+    case _ => false
+  }
+
+  override def hashCode(): Int = rootPaths.toSet.hashCode()
+
+  /**
+   * recursively lists only the partition dirs. Uses the number of partition cols
+   * from user specified schema.
+   * @param path
+   * @param partitionIndex
+   * @return
+   */
+  private def listLeafDirs(path: Path, partitionIndex: Int): mutable.LinkedHashSet[FileStatus] = {
+    val output = mutable.LinkedHashSet[FileStatus]()
+    val numPartitions = userSpecifiedSchema.fields.length
+    if (partitionIndex < numPartitions) {
+      path.getFileSystem(sparkSession.sessionState.newHadoopConf())
```

Comment:
Is there anything better we can do than to make a newconf every time?  (ie, can this be on the class)